sharedbuffers/mapped_struct.py

```diff
@@ -2624,17 +2624,18 @@ def _merge_all(parts, dtype):
         return _merge_all(nparts, dtype)
 
 @cython.cfunc
-@cython.locals(parts = list, discard_duplicate_keys = cython.bint, discard_duplicates = cython.bint)
+@cython.locals(discard_duplicate_keys = cython.bint, discard_duplicates = cython.bint)
 def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplicates):
     if discard_duplicate_keys or discard_duplicates:
         # What numpy.unique does, but using MUCH less RAM
         vpart = apart.view(struct_dt)
         vpart.sort(0)
         if discard_duplicate_keys:
-            flags = numpy.concatenate([[True], apart[1:,0] != apart[:-1,0]])
+            flags = apart[1:,0] != apart[:-1,0]
         elif discard_duplicates:
-            flags = numpy.concatenate([[True], vpart[1:,0] != vpart[:-1,0]])
+            flags = vpart[1:,0] != vpart[:-1,0]
         if not numpy.all(flags):
+            flags = numpy.concatenate([[True], flags])
             apart = apart[flags]
```

Comment:
the name is `discard_duplicates`
but:
```
array([[ 1.,  0.,  0.],
       [ 2.,  3.,  4.],
       [ 1.,  2.,  4.]])
>>> apart
array([[ 1.,  0.,  0.],
       [ 1.,  1.,  2.],
       [ 2.,  3.,  4.],
       [ 1.,  2.,  4.],
       [ 1.,  3.,  5.]])
>>> apart[1:,0] != apart[:-1,0]
array([False,  True,  True, False], dtype=bool)
>>> apart[[numpy.concatenate([[True],(apart[1:,0] != apart[:-1,0])])]]
array([[ 1.,  0.,  0.],
       [ 2.,  3.,  4.],
       [ 1.,  2.,  4.]])
```

Reply:
You're using the `discard_duplicate_keys` code path, which is supposed to deduplicate based on the first column only. So it's correct.

Furthermore, you're skipping the sort pass, which makes the result also incorrect.